File: SMILE/tutorials/cpp/tutorial6.cpp
// tutorial6.cpp
// Tutorial6 creates a dynamic Bayesian network (DBN),
// performs the inference, then saves the model to disk.

#include "smile.h"
#include <cstdio>


static int CreateCptNode(
    DSL_network &net, const char *id,
    const char *name, const char *outcomes[],
    int xPos, int yPos);

static void UpdateAndShowTemporalResults(DSL_network &net);


int Tutorial6()
{
    printf("Starting Tutorial6...\n");

    DSL_errorH().RedirectToFile(stdout);
    
    DSL_network net;

    const char *LOCATION_OUTCOMES[] = { "Pittsburgh", "Sahara", NULL };
    int loc = CreateCptNode(net, "Location", "Location",
        LOCATION_OUTCOMES, 160, 360);

    const char *BOOL_OUTCOMES[] = { "true", "false", NULL };
    int rain = CreateCptNode(net, "Rain", "Rain",
        BOOL_OUTCOMES, 380, 240);

    int umb = CreateCptNode(net, "Umbrella", "Umbrella",
        BOOL_OUTCOMES, 300, 100);

    net.SetTemporalType(rain, dsl_temporalType::dsl_plateNode);
    net.SetTemporalType(umb, dsl_temporalType::dsl_plateNode);

    net.AddArc(loc, rain);
    net.AddTemporalArc(rain, rain, 1);
    net.AddArc(rain, umb);

    DSL_nodeDefinition *rainDef = net.GetNode(rain)->Definition();
    DSL_doubleArray rainProbs(rainDef->GetMatrix()->GetSize());
    rainProbs[0] = 0.7;  // P(Rain=true |Location=Pittsburgh)
    rainProbs[1] = 0.3;  // P(Rain=false|Location=Pittsburgh)
    rainProbs[2] = 0.01; // P(Rain=true |Location=Sahara)
    rainProbs[3] = 0.99; // P(Rain=false|Location=Sahara)
    int res = rainDef->SetDefinition(rainProbs);
    if (DSL_OKAY != res)
    {
        return res;
    }

    rainProbs.SetSize(rainDef->GetTemporalDefinition(1)->GetSize());
    rainProbs[0] = 0.7;   // P(Rain=true |Location=Pittsburgh,Rain[t-1]=true)
    rainProbs[1] = 0.3;   // P(Rain=false|Location=Pittsburgh,Rain[t-1]=true)
    rainProbs[2] = 0.3;   // P(Rain=true |Location=Pittsburgh,Rain[t-1]=false)
    rainProbs[3] = 0.7;   // P(Rain=false|Location=Pittsburgh,Rain[t-1]=false)
    rainProbs[4] = 0.001; // P(Rain=true |Location=Sahara,Rain[t-1]=true)
    rainProbs[5] = 0.999; // P(Rain=false|Location=Sahara,Rain[t-1]=true)
    rainProbs[6] = 0.01;  // P(Rain=true |Location=Sahara,Rain[t-1]=false)
    rainProbs[7] = 0.99;  // P(Rain=false|Location=Sahara,Rain[t-1]=false)
    res = rainDef->SetTemporalDefinition(1, rainProbs);
    if (DSL_OKAY != res)
    {
        return res;
    }

    DSL_nodeDefinition *umbDef = net.GetNode(umb)->Definition();
    DSL_doubleArray umbProbs(umbDef->GetMatrix()->GetSize());
    umbProbs[0] = 0.9; // P(Umbrella=true |Rain=true)
    umbProbs[1] = 0.1; // P(Umbrella=false|Rain=true)
    umbProbs[2] = 0.2; // P(Umbrella=true |Rain=false)
    umbProbs[3] = 0.8; // P(Umbrella=false|Rain=false)
    res = umbDef->SetDefinition(umbProbs);
    if (DSL_OKAY != res)
    {
        return res;
    }

    net.SetNumberOfSlices(5);

    printf("Performing update without evidence.\n");
    UpdateAndShowTemporalResults(net);

    printf("Setting Umbrella[t=1] to true and Umbrella[t=3] to false.\n");
    net.GetNode(umb)->Value()->SetTemporalEvidence(1, 0);
    net.GetNode(umb)->Value()->SetTemporalEvidence(3, 1);
    UpdateAndShowTemporalResults(net);

    res = net.WriteFile("tutorial6.xdsl");
    if (DSL_OKAY != res)
    {
        return res;
    }
    
    printf("Tutorial6 complete: Network written to tutorial6.xdsl\n");
    return DSL_OKAY;
}


static void UpdateAndShowTemporalResults(DSL_network &net)
{
    net.UpdateBeliefs();
    int sliceCount = net.GetNumberOfSlices();
    for (int h = net.GetFirstNode(); h >= 0; h = net.GetNextNode(h))
    {
        if (net.GetTemporalType(h) == dsl_temporalType::dsl_plateNode)
        {
            DSL_node *node = net.GetNode(h);
            int outcomeCount = node->Definition()->GetNumberOfOutcomes();
            printf("Temporal beliefs for %s:\n", node->GetId());
            const DSL_Dmatrix *mtx = node->Value()->GetMatrix();
            for (int sliceIdx = 0; sliceIdx < sliceCount; sliceIdx++)
            {
                printf("\tt=%d:", sliceIdx);
                for (int i = 0; i < outcomeCount; i++)
                {
                    printf(" %f", (*mtx)[sliceIdx * outcomeCount + i]);
                }
                printf("\n");
            }
        }
    }
    printf("\n");
}


static int CreateCptNode(
    DSL_network &net, const char *id, const char *name,
    const char *outcomes[], int xPos, int yPos)
{
    int handle = net.AddNode(DSL_CPT, id);
    DSL_node *node = net.GetNode(handle);

    DSL_idArray ida;
    for (const char **p = outcomes; *p != NULL; p++)
    {
        ida.Add(*p);
    }
    DSL_nodeDefinition *def = node->Definition();
    def->SetNumberOfOutcomes(ida);

    node->Info().Header().SetName(name);

    DSL_rectangle &position = node->Info().Screen().position;
    position.center_X = xPos;
    position.center_Y = yPos;
    position.width = 85;
    position.height = 55;

    return handle;
}

File: SMILE/tutorials/java/src/tutorials/Tutorial6.java
package tutorials;

import smile.*;

// Tutorial6 creates a dynamic Bayesian network (DBN),
// performs the inference, then saves the model to disk.

public class Tutorial6 {
    public static void run() {
        System.out.println("Starting Tutorial6...");
        Network net = new Network();
        
        int loc = createCptNode(
        	net, "Location", "Location",
        	new String[] { "Pittsburgh", "Sahara" }, 
        	160, 360);
        
        int rain = createCptNode(
        	net, "Rain", "Rain",
        	new String[] { "true", "false" }, 
        	380, 240);
        
        int umb = createCptNode(
        	net, "Umbrella", "Umbrella",
        	new String[] { "true", "false" },
        	300, 100);
        
        net.setNodeTemporalType(rain, Network.NodeTemporalType.PLATE);
        net.setNodeTemporalType(umb, Network.NodeTemporalType.PLATE);
        
        net.addArc(loc, rain);
        net.addTemporalArc(rain, rain, 1);
        net.addArc(rain, umb);
    
        double[] rainDef = new double[] {
            0.7,  // P(Rain=true |Location=Pittsburgh)
            0.3,  // P(Rain=false|Location=Pittsburgh)
            0.01, // P(Rain=true |Location=Sahara)
            0.99  // P(Rain=false|Location=Sahara)
        };
        net.setNodeDefinition(rain, rainDef);

        double[] rainDefTemporal = new double[] {
            0.7,   // P(Rain=true |Location=Pittsburgh,Rain[t-1]=true)
            0.3,   // P(Rain=false|Location=Pittsburgh,Rain[t-1]=true)
            0.3,   // P(Rain=true |Location=Pittsburgh,Rain[t-1]=false)
            0.7,   // P(Rain=false|Location=Pittsburgh,Rain[t-1]=false)
            0.001, // P(Rain=true |Location=Sahara,Rain[t-1]=true)
            0.999, // P(Rain=false|Location=Sahara,Rain[t-1]=true)
            0.01,  // P(Rain=true |Location=Sahara,Rain[t-1]=false)
            0.99  // P(Rain=false|Location=Sahara,Rain[t-1]=false)
		};
        net.setNodeTemporalDefinition(rain, 1, rainDefTemporal);
        
        double[] umbDef = new double[] {
    		0.9, // P(Umbrella=true |Rain=true)
            0.1, // P(Umbrella=false|Rain=true)
            0.2, // P(Umbrella=true |Rain=false)
            0.8  // P(Umbrella=false|Rain=false)		
        };
        net.setNodeDefinition(umb, umbDef);
        
        net.setSliceCount(5);

        System.out.println("Performing update without evidence.");
        updateAndShowTemporalResults(net);

        System.out.println(
            "Setting Umbrella[t=1] to true and Umbrella[t=3] to false.");
        net.setTemporalEvidence(umb, 1, 0);
        net.setTemporalEvidence(umb, 3, 1);
        updateAndShowTemporalResults(net);

        net.writeFile("tutorial6.xdsl");
        System.out.println(
            "Tutorial6 complete: Network written to tutorial6.xdsl");
    }
    
    
    private static void updateAndShowTemporalResults(Network net) {
    	net.updateBeliefs();
	    int sliceCount = net.getSliceCount();
	    for (int h = net.getFirstNode(); h >= 0; h = net.getNextNode(h)) {
	    	if (net.getNodeTemporalType(h) == Network.NodeTemporalType.PLATE) {
	    		int outcomeCount = net.getOutcomeCount(h);
	            System.out.printf(
                    "Temporal beliefs for %s:\n", net.getNodeId(h));
	            double[] v = net.getNodeValue(h);
				for (int sliceIdx = 0; sliceIdx < sliceCount; sliceIdx++) {
					System.out.printf("\tt=%d:", sliceIdx);
					for (int i = 0; i < outcomeCount; i++) {
					    System.out.printf(" %f", v[sliceIdx*outcomeCount+i]);
					}
					System.out.println();
				}
	        }
	    }
	    System.out.println();
    }
    
    
    private static int createCptNode(
            Network net, String id, String name, 
            String[] outcomes, int xPos, int yPos) {
        int handle = net.addNode(Network.NodeType.CPT, id);
        
        net.setNodeName(handle, name);
        net.setNodePosition(handle, xPos, yPos, 85, 55);
        
        int initialOutcomeCount = net.getOutcomeCount(handle); 
        for (int i = 0; i < initialOutcomeCount; i ++) {
            net.setOutcomeId(handle, i, outcomes[i]);
        }
        
        for (int i = initialOutcomeCount; i < outcomes.length; i ++) {
            net.addOutcome(handle, outcomes[i]);
        }
        
        return handle;
    }

}
